Track and emit segment loading rate for HttpLoadQueuePeon on Coordinator

kfaraz opened the pull request:
### Design

The loading rate is computed as a moving average of at least the last 10 GiB of successful segment loads.

To account for multiple loading threads on a server, we use the concept of a batch to track load times.
A batch is a set of segments added by the coordinator to the load queue of a server in one go.

```java
batchDurationMillis = t(load queue becomes empty) - t(first load request in batch is sent to server)

batchBytes = total bytes successfully loaded in batch

avg loading rate in batch (kbps) = (8 * batchBytes) / batchDurationMillis

overall avg loading rate (kbps) = (8 * sumOverWindow(batchBytes)) / sumOverWindow(batchDurationMillis)
```

### Changes

- Add `LoadingRateTracker` which computes a moving average load rate based on
the last few GBs of successful segment loads.
- Emit metric `segment/loading/rateKbps` from the Coordinator. In the future, we may
also consider emitting this metric from the historicals themselves.
- Add `expectedLoadTimeMillis` to response of API `/druid/coordinator/v1/loadQueue?simple`

### Testing done
- Added new unit tests
- Performed testing on local cluster with ~1M segments

### Release note
- Emit new metric `segment/loading/rateKbps` from the Coordinator which tracks
the current segment loading rate of a server. Dimensions: `server`
- Add `expectedLoadTimeMillis` to response of API `/druid/coordinator/v1/loadQueue?simple`

### Future work

The `expectedLoadTimeMillis` can be used to show expected time on the web-console.

__This PR does not include web-console changes and the screenshot below is only for reference.__

<img width="944" alt="load_time" src="https://github.com/apache/druid/assets/18635897/e3f4716b-c74e-4161-bfc1-15d1835a92e2">

---

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/http/CoordinatorResource.java:
```diff
@@ -111,14 +111,23 @@ public Response getLoadQueue(
                     .put("segmentsToDropSize", dropSize)
+                    .put("expectedLoadTimeMillis", expectedLoadTimeMillis)
                     .build();
               }
```
Curious why some of these stats are only available in the `simple` mode. I see with `full`, the response being returned currently is less flexible to add new things in this code...

kfaraz replied:
Actually, the `full` mode is currently broken. It just returns the entire `LoadQueuePeon` object in the response.
`CuratorLoadQueuePeon` had some `@JsonProperty` fields that were serialized out into the response.
But `HttpLoadQueuePeon` has never had any `@JsonProperty` fields since it was first written, so we just get an empty object in the response of the API `/druid/v1/coordinator/loadQueue?full`.

Since no one has ever reported this issue, I assume no one is using it (the web-console certainly doesn't use it).
I will create a separate PR to either fix it or just get rid of it completely.

abhishekrb19 commented on server/src/test/java/org/apache/druid/server/http/CoordinatorResourceTest.java:
```diff
@@ -73,4 +74,29 @@ public void testIsLeader()
+  {
+    EasyMock.expect(mock.getLoadManagementPeons())
+            .andReturn(ImmutableMap.of("hist1", new TestLoadQueuePeon()))
+            .anyTimes();
```
```suggestion
            .once();
```

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadingRateTracker.java:
```diff
@@ -0,0 +1,118 @@
+        updatedTotal.increment(-evictedHead.bytes, -evictedHead.millisElapsed);
+      }
+
+      if (updatedTotal.bytes > 0 && updatedTotal.millisElapsed > 0) {
```
Redundant check that will always be true given we check for input `bytes` and `millisElapsed` in line 46

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadingRateTracker.java:
```diff
@@ -0,0 +1,118 @@
+
+  public synchronized void updateProgress(long bytes, long millisElapsed)
+  {
+    if (bytes >= 0 && millisElapsed > 0) {
```
Should this be `bytes > 0`?

kfaraz replied:
Yeah, I think that would be fair.

Although, I am not sure about tombstone segments. They would have size zero but it would still take non-zero time to load time. I guess in that case it would make sense to account for the time in the overall window average.

Let me know what you think.

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadingRateTracker.java:
```diff
@@ -0,0 +1,118 @@
+ * <p>
+ * The loading rate is computed as a moving average of the last
+ * {@link #MOVING_AVERAGE_WINDOW_SIZE} progress updates (or more if any of the
+ * updates was smaller than {@link #MIN_ENTRY_SIZE_BYTES}).
```
```suggestion
 * updates were smaller than {@link #MIN_ENTRY_SIZE_BYTES}).
```

abhishekrb19 reviewed (approved):
Nice! Looks good to me. I left a few small suggestions and questions (mostly out of curiosity).

abhishekrb19 commented on docs/operations/metrics.md:
```diff
@@ -325,6 +325,7 @@ These metrics are for the Druid Coordinator and are reset each time the Coordina
 |`segment/dropSkipped/count`|Number of segments that could not be dropped from any server.|`dataSource`, `tier`, `description`|Varies|
 |`segment/loadQueue/size`|Size in bytes of segments to load.|`server`|Varies|
 |`segment/loadQueue/count`|Number of segments to load.|`server`|Varies|
+|`segment/loading/rateKbps`|Moving average rate of segment loading on a server in kbps.|`server`|Varies|
```
Looks like spellcheck is failing because of `kbps`. The spelling seems valid, so we can ignore it by adding it to the `.spelling` file.

abhishekrb19 replied:
Should we also document the window constraints (window size and bytes) in the user-facing docs? I think mentioning it will help users better interpret this moving average metric.

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadingRateTracker.java:
```diff
@@ -0,0 +1,118 @@
+
+  public synchronized void updateProgress(long bytes, long millisElapsed)
+  {
+    if (bytes >= 0 && millisElapsed > 0) {
```
```suggestion
    if (bytes > 0 && millisElapsed > 0) {
```

kfaraz commented on server/src/main/java/org/apache/druid/server/coordinator/loading/LoadingRateTracker.java:
```diff
@@ -0,0 +1,118 @@
+        updatedTotal.increment(-evictedHead.bytes, -evictedHead.millisElapsed);
+      }
+
+      if (updatedTotal.bytes > 0 && updatedTotal.millisElapsed > 0) {
+        windowTotal.set(updatedTotal);
+      }
```
```suggestion
      windowTotal.set(updatedTotal);
```

kfaraz commented:
@abhishekrb19 , just realized that there is actually an error in the logic to compute the time taken to load. 😛 

Imagine the following scenario:
- t0: Coordinator wants to load 4 segments on Historical.
- t0: Coordinator sends request to Historical to load the 4 segments.
- t1: Historical sends back a response as soon as any one segment finishes loading.
- t1: Coordinator measures time spent in the request = `(t1 - t0)`.
- t2: Coordinator sends next request to historical. This request again contains the remaining 3 segments which the Coordinator thinks are not loaded yet.
- t3: Historical sends back a response saying that the remaining 3 segments have been loaded.
- t3: Coordinator measures time spent in second request  = `(t3 - t2)`

So total time measured by Coordinator = `(t3 - t2) + (t1 - t0)`
But actual total time taken by Historical to load all segments = `(t3 - t0)`

I am trying to figure out how to fix this. Will update once I have a solution.

abhishekrb19 commented on server/src/main/java/org/apache/druid/server/coordinator/loading/HttpLoadQueuePeon.java:
```diff
@@ -542,6 +557,24 @@ private void onRequestCompleted(SegmentHolder holder, RequestStatus status)
+        responses.stream()
+                 .filter(response -> response.getStatus().getState() == SegmentChangeStatus.State.SUCCESS)
+                 .map(DataSegmentChangeResponse::getRequest)
+                 .filter(req -> req instanceof SegmentChangeRequestLoad)
```
Makes sense, thanks for the explanation

kfaraz commented:
Update: Logic is fixed, need to test it out a little more.

Once that is done, I will update the description and documentation here.

kfaraz commented:
@abhishekrb19 , I have updated the docs, description in the PR.
Please let me know what you think.

AmatyaAvadhanula commented:
Could you please share the motivation behind these changes?
Would a simpler approach such as [sum(segment size) / time] across successful loads in a coordinator cycle not be sufficient?

kfaraz commented:
Thanks for the feedback, @AmatyaAvadhanula !

>  Would a simpler approach such as [sum(segment size) / time] across successful loads in a coordinator cycle not be sufficient?

Segment loads are not tied to a coordinator cycle.

"Coordinator cycle" or "coordinator run" simply refers to a single invocation of a duty
like `RunRules` or `BalanceSegments`. After the duty has run and assigned a bunch
of segments to the load queue, the segments may take any amount of time to finish loading.

While summing up the sizes of successfully loaded segments is trivial,
the definition of _time elapsed_ is what complicates the whole logic.

Problems:
1. We want some kind of a moving average.
2. Segments assigned in one coordinator run may remain in the queue for several runs.
So when is the start time and end time?
3. While there are already segments in queue, the next coordinator run may assign more segments.
How would this affect start time and end time?
---
The simplest (and most intuitive) thing to do would be to track the load time of each segment
individually. I actually started out doing this. Start time would be the time when the request to
load that segment is first sent to the server. End time would be when the request succeeds.

This design alternative has been alluded to in the PR description as well.

__But this would be incorrect,__ since while a segment is being loaded on the historical by one thread,
another thread could be loading another segment.

In other words, _the segment load durations are not mutually exclusive,_ so we can't simply sum them up.
If we did, the computed loading rate would be lower than the actual (not the end of the world but still).

That said, if there is only one loading thread on the server (which is often the case), then the naive logic works just fine.
```
numLoadingThreads = Math.max(1, JvmUtils.getRuntimeInfo().getAvailableProcessors() / 6)
```

---

Let me know what you think.
If you feel this seems too complicated and we could get away with the naive logic for now,
I can just do that and save this convoluted design for a rainy day 😂 . Once we have seen
the feature in action, we will know for sure. In the future, if the Coordinator could know the number
of loading threads on the server, we could just multiply the computed rate by num threads to offset
the effect of summing up the times.

cc: @abhishekrb19

kfaraz commented:
Reverted to simpler logic that works well for single threads.

cc: @AmatyaAvadhanula , @abhishekrb19

abhishekrb19 commented:
@kfaraz, thanks for the updates, I haven't looked into the latest changes yet. But a quick question based on the comments:

>Reverted to simpler logic that works well for single threads.

Does this mean that the simpler logic now only accounts for the default case of a single loading thread (`druid.segmentCache.numLoadingThreads`)? That seems reasonable for now if we don't want the added complexity, but what is reported when there are multiple loading threads? Instead of reporting potentially incorrect values, I think we can skip reporting the metric and including `expectedLoadTimeMillis` in the API until we have proper support for accounting for multiple threads.

kfaraz commented:
> Does this mean that the simpler logic now only accounts for the default case of a single loading thread (druid.segmentCache.numLoadingThreads)?

Yes, @abhishekrb19 .

> Instead of including potentially incorrect values, I think we can skip reporting the metric and including expectedLoadTimeMillis in the API for the multiple threads case until we have proper support for it.

Currently, there is no way for the coordinator to know the number of loading threads being used by a historical.
Also, I think it is fine to report somewhat incorrect values as this is an experimental feature now anyway.
It would be interesting to see how off the values turn out to be when using multiple loading threads.
Moreover, we would be over-estimating the expected time, never under-estimating it.
It's always a good feeling when downloads take less time than originally expected! 😛

abhishekrb19 commented:
Got it. Yeah, making it an experimental feature makes sense. We can also call out the caveat about multiple loading threads in the docs. I will try to look at the latest updates soon!

abhishekrb19 commented:
@kfaraz, apologies for the delay in getting back. 

The docs recommend having at least 16 [vCPUs](https://druid.apache.org/docs/latest/tutorials/cluster/#data-server) for data servers, so there will be at least 2 loading threads by default in production clusters. As to how much overlap there is between the time spent by loading threads, I'm not sure. Here are a few exploratory thoughts/ideas to simplify and track this more accurately:

1. How about tracking the load rate directly in the historicals/data servers? I see you have listed that as a potential approach for the future. Besides it being useful and more accurate, I think it's also relatively straightforward to implement. Given that the `SegmentLoadDropHandler` code is already processing batches of change requests, I think we piggyback on the logic to add some tracking there. Also, we don't introduce another notion of "batch" in the coordinator's `HttpLoadQueuePeon` if we decide to revive that idea. 

I think one downside to this is that the rate computed on the historicals won't account for the end-to-end time (e.g., time spent in the queue, etc). If that is significant, we can perhaps track a metric separately on the coordinator?

2. If we want to compute the aggregated rate from the coordinator, we could perhaps expose an internal API on the historicals that the coordinator can then query to get the requested info (# of loading threads, # of segment batches processed, etc.) if they're available. However, I think this approach might be an overkill.

Please let me know what you think.

kfaraz commented:
@abhishekrb19 , based on our offline discussion, I have updated the PR.
Please let me know what you think.

abhishekrb19 reviewed (approved):
Looks good overall to me 👍. Thanks @kfaraz!

kfaraz commented:
Thanks a lot for the feedback, @abhishekrb19 ! 😃 
I have updated some javadocs, added some new ones and renamed the `reset()` to `stop()` as suggested.
